clash-cores/src/Clash/Cores/Xilinx/Floating/Explicit.hs

```diff
@@ -230,6 +234,43 @@ div = withFrozenCallStack $ divWith defConfig
 -- | The default delay for floating point division with default customization.
 type DivDefDelay = 28
 
+-- | Customizable conversion of @Unsigned 32@ to @Float@
+--
+-- Only the delay is configurable, so this function does not take a @Config@
+-- argument.
+fromU32With
+  :: forall d dom n
+   . ( KnownDomain dom
+     , KnownNat d
+     , HasCallStack
+     )
+  => Clock dom
+  -> Enable dom
+  -> DSignal dom n (Unsigned 32)
+  -> DSignal dom (n + d) Float
+fromU32With clk en = delayI und en clk . fmap fromIntegral
+ where
+  und = withFrozenCallStack $ errorX "Initial values of fromU32 undefined"
+{-# NOINLINE fromU32With #-}
+{-# ANN fromU32With (vhdlFromUPrim 'fromU32With "fromU32") #-}
+{-# ANN fromU32With (veriFromUPrim 'fromU32With "fromU32") #-}
+
+-- | Conversion of @Unsigned 32@ to @Float@, with default delay
+fromU32
+  :: forall dom n
+   . ( KnownDomain dom
+     , HasCallStack
+     )
+  => Clock dom
+  -> Enable dom
+  -> DSignal dom n (Unsigned 32)
+  -> DSignal dom (n + FromU32DefDelay) Float
+fromU32 = withFrozenCallStack fromU32With
+{-# INLINE fromU32 #-}
+
+-- | The default delay for conversion of @Unsigned 32@ to @Float@
```

Comment:
Could you comment whether this optimizes for anything? Clock speed I suppose?

Reply:
Xilinx wrote a whole section on it:
> Latency and Rate Configuration
This parameter describes the number of cycles between an operand input and result output.
The latency of all operators can be set between 0 and a maximum value that is dependent
upon the parameters chosen. IP cores that are intended for many applications, such as the
Floating-Point Operator core, are designed to run as fast as the DSP primitives. To achieve
this, the fabric logic must be heavily pipelined, leading to relatively high latency. For
designs which run at a relatively low frequency, the latency can be reduced while timing can
still be met. For instance, if the fully pipelined design with latency=12 can meet timing at
400 MHz, then if the system clock in the user design is 70MHz, for example, then latency
can likely be reduced to six and timing can still be met. However, the relationship between
latency and achieveable clock speed is not linear. This is because the amount of logic
between register stages is roughly the same, so if one register is removed, the achieveable
clock frequency drops considerably because one register to register path now has almost
double the amount of logic as other register-to-register delays. The advantage of reducing
latency when clock speed allows is that the result appears sooner. There is no significant
gain in resources other than an inevitable reduction in registers for which there is an ample
supply. For the previous example, Xilinx recommends using latency=12 if the system clock
is > 0.5 * value in the resource and performance section on the Xilinx webpage for the IP
core. Other latency values of use are 6, 3, 2 and 1, though minimum may apply due to FIFOs
and feedback stages. Therefore, there is not much use in setting latency to 11 for this
reason.
>
> Note: The Accumulator operator has a minimum latency of 1.

The only things I wrote on it in this module's doc is:
> Most functions allow customization of the Xilinx IP. Valid combinations will need to be gleaned from, e.g., the Vivado wizard (IP Catalog -> Floating-point). 

> For each customizable operation, there also exists a function that uses the defaults. These functions use the settings from `defConfig` and the maximum delay for the Xilinx IP with that configuration. That delay is also defined as a type variable for delay annotation in circuits.

To answer your question briefly, yes, I think we can conclude from what Xilinx writes that this optimises for clock speed. Even though in this specific case, there is no DSP involved. Also, to answer more extensively, they appear to say you should only divide by integers for useful results, but their example is very simple: they pick 12, and useful latencies are clearly 6, 3, 2 and 1. But how to do that with 5? Is it 3, 2, 1 or 2, 1? Your guess is as good as mine. If you _really_ wanted to know, I suppose you could determine Fmax for 5, 3 and 2. I don't want to know that badly, though.

Reply:
Nah it's fine!

Reply:
Thanks for the comment!